### ISSUE TITLE
On databases that have a default charset that is not utf-8, data can get corrupted

### ISSUE THREAD

**Issue by ataraxie** [piwik/piwik#6497]
Matt suggested me to create a bug report for this problem which was discussed in detail in this forum thread:
http://forum.piwik.org/read.php?2,120225,120539

### Scenario
As Piwik admin user, navigate to Settings => Websites. I always get "You currently have access to 0 website". This happens when I am superuser as well as when I have dedicated "admin" permission to websites.

### Problem source
* The JSON response of the request is empty with Content-Length: 0
* Request: http://mypiwikhost/index.php?date=yesterday&fetchAliasUrls=true&format=JSON2&idSite=1&method=SitesManager.getSitesWithAdminAccess&module=API&period=day&token_auth=XXXXXXXXX
* This is the case as soon as at least one Piwik site has an umlaut ("ä", "ü", "ö" ...) in its name
* If format=XML is set in the request, it works fine

### Environment
* PHP 5.5.10
* Internet Information Services 7.5 (IIS)
* MySQL 5.5.27
* Windows Server 2008 R2

Since Piwik allows umlauts for site name, I consider this a Piwik bug. We have an enterprise customer installation of Piwik running with a database of ~ 40 GB. Our customer is very upset with this at the moment, since, with regard to the time I needed to figure out the problem source (see forum thread), this issue has kept us from updating from 1.12 to 2.x for several months. So I kindly ask you to consider this a major issue. Also, as the forum thread shows, I'm not the only one affected.

**Comment by mattab** [comment]
Again today on the forums someone reported that adding charset=utf8 solved a problem for them. Maybe we should investigate this and really understand the problem. Or maybe we should simply add charset=utf8 to the config by default, to prevent such issue all together?

**Comment by mattab** [comment]
The goal of this issue is to estimate whether there is any performance overhead when we add `charset = utf8` in the global config file by default. This will result in running `SET NAMES` query on each request - is there is no overhead doing this, let's add it for sure as it make Piwik work well across more DB configs!

(because we told users of the workaround `charset=utf8` many times in the forums in the past, this little change would help much. added `Performance` tag and renamed issue)

**Comment by tsteur** [comment]
It takes on my instance about 60 microsecs with Mysqli and 100 microsecs with PDO so should not cause any problems. Not sure how it behaves under load but don't think that it causes any problems.

Maybe the `Tracker\DB` could be improved as it does a separate `set names` query https://github.com/piwik/piwik/blob/2.13.0-b3/core/Tracker/Db/Pdo/Mysql.php#L92-L95 whereas in normal `DB` it is handled directly in DSN and driver option https://github.com/piwik/piwik/blob/2.13.0-b3/libs/Zend/Db/Adapter/Pdo/Mysql.php#L85-L107

I presume second version (dsl + init command) should be even faster

**Comment by mattab** [comment]
> Maybe the Tracker\DB could be improved as it does a separate set names query  whereas in normal DB [...]

reading your comment i remembered we wanted to consider: Unify Tracker/Db and Db into one? #7727

**Comment by saleemkce** [comment]
Yes, I have had bad experiences in the past in OSIpage app where users would post lots of URLs from different languages. We had some key tables where URLs were stored were using default latin1 charset which seemed to garble URLs.

The fix had been quite simple. We need to make sure that, 
* when creating database, we should use database collation as either "utf8_general_ci" or "utf8_unicode_ci"
     -> **utf8_general_ci** is a little bit faster than **utf8_unicode_ci**
     -> **utf8_unicode_ci** is more accurate than **utf8_general_ci** according to myql documentation.
     -> the best bet may be using "utf8_unicode_ci" as we would always prefer accuracy over small performance improvement.

* all tables should use **charset as utf8** and **collate as utf8_unicode_ci**   

In other words, we could use this mysql command to change a table's charset one at a time.

````
ALTER TABLE myTable CHARACTER SET utf8 COLLATE utf8_unicode_ci;
````

* any website displaying content other than English, should have this in HTML head tag so that it handles other language characters  properly & correctly on web page.

````
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">
````

After applying these changes in OSIpage app, we didn't ever come across URL garble issues again. I hope this solves the issue.

**Comment by mattab** [comment]
> It takes on my instance about 60 microsecs with Mysqli and 100 microsecs with PDO so should not cause any problems. Not sure how it behaves under load but don't think that it causes any problems.

@tsteur Awesome! moving into 2.15.0

**Comment by yashweb3** [comment]
I found your solution here in GitHub for this kind of Problem: adding the line: charset = utf8 in the database section of config.ini.php.

I did this as well and the websites all reappeared in the settings.

But when I now look at my data the page titels are not displayed correct and look like this:

BroschÃ¼ren --> should be: Broschüren
SpaÃŸbÃ¤der --> should be: Spaßbäder

The database is utf-8 and the pages are all utf-8 as well.

What can I do?

Best regards

Christoph

>>i am getting same problem what was the solution :)

**Comment by yashweb3** [comment]
> I found your solution here in GitHub for this kind of Problem: adding the line: charset = utf8 in the database section of config.ini.php.
> 
> I did this as well and the websites all reappeared in the settings.
> 
> But when I now look at my data the page titels are not displayed correct and look like this:
> 
> BroschÃ¼ren --> should be: Broschüren
> SpaÃŸbÃ¤der --> should be: Spaßbäder
> 
> The database is utf-8 and the pages are all utf-8 as well.
> 
> What can I do?
> 
> Best regards
> 
> Christoph

Did you find any solution for it?

**Comment by matthewhutchings** [comment]
This is still an issue.

**Comment by mattab** [comment]
I'm not sure if it's helpful, but this is the FAQ we have that explains how to convert a Matomo DB to utf8mb4 charset, which is what we've been using by default https://matomo.org/faq/how-to-update/how-to-convert-the-database-to-utf8mb4-charset/
